File: pyskynet/foreign.py

import pyskynet.skynet_py_foreign_seri as foreign_seri
import pyskynet.skynet as skynet
import pyskynet.skynet_py_mq as skynet_py_mq


PTYPE_FOREIGN = skynet_py_mq.PTYPE_FOREIGN
PTYPE_FOREIGN_REMOTE = skynet_py_mq.PTYPE_FOREIGN_REMOTE


class CMDDispatcher(object):
    def __init__(self):
        self.cmd2func = {}

    def __call__(self, first, *args):
        return self.cmd2func[first](*args)

    def __setitem__(self, k, v):
        if isinstance(k, str):
            self.cmd2func[k] = v
            self.cmd2func[k.encode("ascii")] = v
        elif isinstance(k, bytes):
            self.cmd2func[k] = v
            self.cmd2func[k.decode("ascii")] = v
        else:
            self.cmd2func[k] = v

    def __getitem__(self, k):
        return self.cmd2func[k]


CMD = CMDDispatcher()

remotepack = foreign_seri.remotepack
remoteunpack = foreign_seri.remoteunpack

trash = foreign_seri.trash


def __foreign_dispatch(session, source, argtuple):
    ret = CMD(*argtuple)
    if session != 0:
        if type(ret) == tuple:
            msg_ptr, msg_size = foreign_seri.__refpack(*ret)
        else:
            msg_ptr, msg_size = foreign_seri.__refpack(ret)
        hook_ptr = foreign_seri.__packhook(msg_ptr)
        if hook_ptr:
            if not skynet.ret(hook_ptr, msg_size):
                foreign_seri.__unref(msg_ptr)
                foreign_seri.trash(msg_ptr)
        else:
            skynet.ret(msg_ptr, msg_size)


def __foreign_remote_dispatch(session, source, argtuple):
    ret = CMD(*argtuple)
    if session != 0:
        if type(ret) == tuple:
            skynet.ret(*foreign_seri.remotepack(*ret))
        else:
            skynet.ret(*foreign_seri.remotepack(ret))

def __dontpackhere():
    raise skynet.PySkynetCallException("don't use pack here")

# dispatch foreign message
skynet.register_protocol(
        id=PTYPE_FOREIGN,
        name="foreign",
        pack=__dontpackhere,
        unpack=foreign_seri.__refunpack,
        dispatch=__foreign_dispatch,
        )

# dispatch foreign message
skynet.register_protocol(
        id=PTYPE_FOREIGN_REMOTE,
        name="foreign_remote",
        pack=__dontpackhere,
        unpack=foreign_seri.remoteunpack,
        dispatch=__foreign_remote_dispatch,
        )


def dispatch(cmd=None, func=None):
    global CMD

    def wrapper(func):
        CMD[cmd] = func
        return func
    if not (func is None):
        assert callable(func), "dispatch's second arg must be callable"
        CMD[cmd] = func
    elif callable(cmd):
        CMD = func
    elif isinstance(cmd, dict):
        for k, v in cmd.items():
            CMD[k] = v
    elif isinstance(cmd, str) or isinstance(cmd, bytes):
        return wrapper
    else:
        raise Exception("dispatch failed for unexception args")

def __safe_rawsend(addr, zero_none, msg_ptr, msg_size):
    hook_ptr = foreign_seri.__packhook(msg_ptr)
    if hook_ptr:
        session = skynet_py_mq.csend(addr, PTYPE_FOREIGN, zero_none, hook_ptr, msg_size)
        if session is None:
            foreign_seri.__unref(msg_ptr)
            foreign_seri.trash(msg_ptr)
    else:
        session = skynet_py_mq.csend(addr, PTYPE_FOREIGN, zero_none, msg_ptr, msg_size)
    return session


def call(addr, *args):
    msg_ptr, msg_size = foreign_seri.__refpack(*args)
    session = __safe_rawsend(addr, None, msg_ptr, msg_size)
    if session is None:
        raise psproto.PySkynetCallException("send to invalid address %08x" % dst)
    re = skynet.__yield_call(addr, session)
    return foreign_seri.__refunpack(*re)


def send(addr, *args):
    msg_ptr, msg_size = foreign_seri.__refpack(*args)
    session = __safe_rawsend(addr, 0, msg_ptr, msg_size)
    return session is not None

File: pyskynet/lualib/pyskynet/foreign.lua

local foreign_seri = require("pyskynet.foreign_seri")
local numsky = require("numsky")

local skynet = require("skynet")
local csend = (require "skynet.core").send

collectgarbage("setpause", 100)
collectgarbage("setstepmul", 200)

local foreign = {}

local PTYPE_FOREIGN_REMOTE = 254
local PTYPE_FOREIGN = 255

foreign.PTYPE_FOREIGN_REMOTE = PTYPE_FOREIGN_REMOTE
foreign.PTYPE_FOREIGN = PTYPE_FOREIGN

foreign.remotepack = assert(foreign_seri.remotepack)
foreign.remoteunpack = assert(foreign_seri.remoteunpack)
foreign.trash = assert(foreign_seri.trash)

foreign.CMD = setmetatable({}, {
	__call=function(t, first, ...)
		local f = t[first]
		if not f then
			error("cmd "..tostring(first).." not found")
		end
		return f(...)
	end
})

local function __foreign_dispatch(session, source, ...)
	if session ~= 0 then
        local msg_ptr, msg_size = foreign_seri.__refpack(foreign.CMD(...))
		local hook_ptr = foreign_seri.__packhook(msg_ptr)
		if hook_ptr then
			if not skynet.ret(hook_ptr, msg_size) then
				foreign_seri.__unref(msg_ptr)
				foreign_seri.trash(msg_ptr)
			end
		else
			skynet.ret(msg_ptr, msg_size)
		end
	else
		foreign.CMD(...)
	end
end

local function __foreign_remote_dispatch(session, source, ...)
	if session ~= 0 then
        local msg_ptr, msg_size = foreign_seri.remotepack(foreign.CMD(...))
		skynet.ret(msg_ptr, msg_size)
     else
		foreign.CMD(...)
	end
end

do
	local REG = skynet.register_protocol

	REG {
		name = "foreign",
		id = PTYPE_FOREIGN,
		pack = function()
            error("use foreign.someapi(xxx, ...) instead of skynet.someapi(xxx, 'foreign', ...) when packing foreign message")
        end,
		unpack = foreign_seri.__refunpack,
		dispatch = __foreign_dispatch,
	}

	REG {
		name = "foreign_remote",
		id = PTYPE_FOREIGN_REMOTE,
		pack = function()
            error("foreign_remote pack is not recommand to used here")
        end,
		unpack = foreign_seri.remoteunpack,
		dispatch = __foreign_remote_dispatch,
	}

end

function foreign.dispatch(cmd, func)
	if func then
		foreign.CMD[cmd] = func
	elseif type(cmd) == "table" then
		for k, v in pairs(cmd) do
			foreign.CMD[k] = v
		end
	elseif type(cmd) == "function" then
		foreign.CMD = cmd
	else
		error("dispatch failed for args")
	end
end

local i = 1
while debug.getupvalue(skynet.rawcall, i) ~= "yield_call" do
    i=i+1
end
local _, yield_call = debug.getupvalue(skynet.rawcall, i)

local function safe_rawsend(addr, zero_nil, msg_ptr, msg_size)
	local hook_ptr = foreign_seri.__packhook(msg_ptr)
	local session
	if hook_ptr then
		session = csend(addr, PTYPE_FOREIGN , zero_nil, hook_ptr, msg_size)
		if session == nil then
			foreign_seri.__unref(msg_ptr)
			foreign_seri.trash(msg_ptr)
		end
	else
		session = csend(addr, PTYPE_FOREIGN , zero_nil, msg_ptr, msg_size)
	end
	return session
end

function foreign.call(addr, ...)
    local msg_ptr, msg_size = foreign_seri.__refpack(...)
	local session = safe_rawsend(addr, nil, msg_ptr, msg_size)
	if session == nil then
		error("call to invalid address " .. skynet.address(addr))
	end
	return foreign_seri.__refunpack(yield_call(addr, session))
end

function foreign.send(addr, ...)
    local msg_ptr, msg_size = foreign_seri.__refpack(...)
	return safe_rawsend(addr, 0, msg_ptr, msg_size)
end

return foreign
